applications/rollout_ensemble.py

```diff
@@ -0,0 +1,626 @@
+# ---------- #
+# System
+import os
+import gc
+import sys
+import yaml
+import logging
+import warnings
+import multiprocessing as mp
+from pathlib import Path
+from argparse import ArgumentParser
+from collections import defaultdict
+
+# ---------- #
+# Numerics
+from datetime import datetime, timedelta
+import pandas as pd
+import xarray as xr
+import numpy as np
+
+# ---------- #
+import torch
+
+# ---------- #
+# credit
+from credit.models import load_model
+from credit.seed import seed_everything
+from credit.data import (
+    concat_and_reshape,
+    reshape_only
+)
+from credit.datasets import setup_data_loading
+from credit.transforms import load_transforms, Normalize_ERA5_and_Forcing
+from credit.pbs import launch_script, launch_script_mpi
+from credit.pol_lapdiff_filt import Diffusion_and_Pole_Filter
+from credit.metrics import LatWeightedMetrics, LatWeightedMetricsClimatology
+from credit.forecast import load_forecasts
+from credit.distributed import distributed_model_wrapper, setup, get_rank_info
+from credit.models.checkpoint import load_model_state, load_state_dict_error_handler
+from credit.postblock import GlobalMassFixer, GlobalWaterFixer, GlobalEnergyFixer
+from credit.parser import credit_main_parser, predict_data_check
+from credit.datasets.era5_predict_batcher import (
+    BatchForecastLenDataLoader,
+    Predict_Dataset_Batcher
+)
+
+
+logger = logging.getLogger(__name__)
+warnings.filterwarnings("ignore")
+os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
+os.environ["OMP_NUM_THREADS"] = "1"
+os.environ["MKL_NUM_THREADS"] = "1"
+
+
+def compute_metrics(metrics, y_pred, y, date_time, forecast_step, utc_datetime):
+    """Compute metrics and update metrics_results."""
+    metrics_results = {}
+    metrics_dict = metrics(y_pred.float(), y.float(), forecast_datetime=date_time)
+    for k, m in metrics_dict.items():
+        metrics_results[k] = m.item()
+    metrics_results["forecast_step"] = forecast_step
+    metrics_results["datetime"] = utc_datetime
+    return metrics_results
+
+
+def predict(rank, world_size, conf, backend=None, p=None):
+    # setup rank and world size for GPU-based rollout
+    if conf["predict"]["mode"] in ["fsdp", "ddp"]:
+        setup(rank, world_size, conf["trainer"]["mode"], backend)
+
+    # infer device id from rank
+    if torch.cuda.is_available():
+        device = torch.device(f"cuda:{rank % torch.cuda.device_count()}")
+        torch.cuda.set_device(rank % torch.cuda.device_count())
+    else:
+        device = torch.device("cpu")
+
+    # config settings
+    seed = conf["seed"]
+    seed_everything(seed)
+
+    # number of input time frames
+    history_len = conf["data"]["history_len"]
+
+    # length of forecast steps
+    lead_time_periods = conf["data"]["lead_time_periods"]
+
+    # batch size
+    batch_size = conf["predict"].get("batch_size", 1)
```

Comment:
I didn't realize `get` was a dict method until your use here just now. This is a really succinct way to handle missing defaults.

Reply:
Yeah super convenient!